src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -716,29 +694,61 @@ PosixEventEngine::GetDNSResolver(
       return ares_resolver.status();
     }
 #if GRPC_ENABLE_FORK_SUPPORT && GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK
-    {
-      grpc_core::MutexLock lock(&resolver_handles_mu_);
-      RegisterResolver(&resolver_handles_,
-                       ares_resolver->get()->GetReinitHandle());
-    }
+    RegisterAresResolverForFork(ares_resolver->get());
 #endif  // GRPC_ENABLE_FORK_SUPPORT && GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK
     return std::make_unique<PosixEventEngine::PosixDNSResolver>(
         std::move(*ares_resolver));
 #else   // GRPC_PLATFORM_SUPPORTS_POSIX_POLLING
     grpc_core::Crash("Can not create CAres resolver with disabled poller");
 #endif  // GRPC_PLATFORM_SUPPORTS_POSIX_POLLING
-#endif  // GRPC_ARES == 1 && defined(GRPC_POSIX_SOCKET_ARES_EV_DRIVER)
   }
   GRPC_TRACE_LOG(event_engine_dns, INFO)
       << "PosixEventEngine::" << this << " creating NativePosixDNSResolver";
   return std::make_unique<NativePosixDNSResolver>(shared_from_this());
```

Comment:
This is now redundant to the NativeResolver version you added below. Previous logic would only create the native resolver in this function if Ares was not supported.

Reply:
Oh, you're right. This code is such a mess. The fact that the code block can be totally empty means the check doesn't need to be outside the macro guards at all. 

```
  if (!ShouldUseAresDnsResolver()) {
    // log something
    return std::make_unique<NativePosixDNSResolver>(shared_from_this());
  }
  #if ...
    // c-ares logic
  #endif
```